### DISCUSSION TITLE
SSG, fallbacks and SEO considerations

### DISCUSSION THREAD

**Discussion by rus-yurchenko** [vercel/next.js#12482] [Help] [answered]
Hi there, I struggle with the following question...

Whether fallback pages will be a blocker for crawlers on their "first visit" since they generate a static HTML with no potential SEO content or they will somehow pass the crawl and can produce correct data for the SEO? It potentially related to OG preview generation and indexing.

What happens from the SEO point when we haven't rendered SSG page yet? How do the not-pre-renreded pages behave for SEO?

**Comment by rafaelalmeidatk** [comment, marked as answer]
Google's crawlers will run JS in the client so the fallbacks won't negatively affect your SEO performance, for others crawlers you can read Joe's response here: https://github.com/zeit/next.js/discussions/11552#discussioncomment-2950

**Reply by rus-yurchenko** [reply]
He didn't clearly reply to what happens with other crawlers. Does it mean that today crawlers will parse pages incorrectly for the first time?

**Reply by rafaelalmeidatk** [reply]
I don't have any info about how other crawlers work so take this with a grain of salt: if the crawler doesn't wait and run JS in the client it might parse the page incorrectly with the content loaders, similar to what happens to client-side-only apps, like create-react-app.

This will not be an issue when they release the blocking page requests for crawlers, with this feature the crawlers will always receive the page with content. @Timer please correct me if I am wrong

**Reply by rus-yurchenko** [reply]
@rafaelalmeidatk yeah, that's my initial question. So currently fallback pages potentially COULD be an issue with crawlers. So... will wait for the blocking page requests feature for crawlers. Thank you for your answer.

**Reply by janus-reith** [reply]
@rafaelalmeidatk Hm, trying around in the google search console I was unable to confirm this, at least for my use case.
With the Live-Test of the Search Console, the first Test on a new fallback url will return an empty loading page.
Subsequent tries will return the proper page which now exists.
I create sitemaps which hint to different valid configurations for a configurable product, and not all of these urls exist initially.

Maybe the Live-Test returns different results than the actual crawler, but so far this unsettles me a little bit.

**Reply by arunoda** [reply]
@Timer is possible to mark this question as Unanswered.
Because, this is an incorrect answer.

Even though Google can read JS (it's a question, it does that for every page), in the case of fallback, it cannot render the page correctly.

In the case of fallback, we are making an HTTP request to get the payload and render the page on the client-side.
I don't think that inside a Google crawler. I didn't check lately, but it didn't work a few years ago.

I also think, selectively blocking fallback is a bad thing.
Let us choose whether we need a blocking fallback or not for a given page.

Is this feature on the roadmap, any PRs?
If not, I can help.

**Reply by timneutkens** [reply]
> In the case of fallback, we are making an HTTP request to get the payload and render the page on the client-side.
I don't think that inside a Google crawler. I didn't check lately, but it didn't work a few years ago.

This works fine, their crawler got significantly more advanced the last year. It also runs the JS indexing significantly faster (few seconds after the http crawl instead of 2 weeks).

> Let us choose whether we need a blocking fallback or not for a given page.

We're planning to add a `fallback: 'block'` option that would have the blocking behavior. Though it's a trade-off as you lose reliably sending a fast response to users around the world

**Reply by arunoda** [reply]
@timneutkens cool.

> We're planning to add a fallback: 'block' option that would have the blocking behavior. Though it's a trade-off as you lose reliably sending a fast response to users around the world

But this only needed for the first request, right?
After the first render, we can use existing SWR behavior.

**Reply by timneutkens** [reply]
> But this only needed for the first request, right?

Correct

**Reply by arunoda** [reply]
Thanks @timneutkens

**Comment by revskill10** [comment]
I have a demo for your purpose [here](https://build.kms-edu1.now.sh/about/charmeleon/nono)
I use a SSR version as fallback. The drawback is that, because we're using now, we have no way to write to the cache at runtime when user first visit a non-static page.
But if you're trying to deploy to your own server, it's possible to do that.

So, when user first visit a non-static page, you will write the page into cache (public path is `/static` to be served from nginx).

My `now.json` is like this
```js
{
  "builds": [
    {
      "src": "node/serverless.js",
      "use": "@now/node",
      "config": {
        "includeFiles": [
          "node/*.json",
          "node/*.js"
        ]
      }
    }, {
      "src": "web/**/*",
      "use": "@now/static"
    }
  ],
  "routes": [
    { "src": "/", "dest": "web/www/index.html" },
    { "src": "/about/wartortle'", "dest": "web/www/about/wartortle.html", "continue": false },
    { "src": "^/static/(.+)", "dest": "web/$1" },
    { "src": "^/(.*)", "dest": "node/serverless.js"}
  ]
}
```
Notice the `/about/wartortle` route, it's a static page !
Hope this will answer your question about this issue.
I hope in future `now` will allow us to control the writable storage system from `static` build to achive the above result.
